parsers/ansibleplaybook.c

```diff
@@ -0,0 +1,233 @@
+		popBlockType (state, token);
+}
+
+static bool stateStackMatch (struct yamlBlockTypeStack *stack,
+								yaml_token_type_t *expectation,
+								unsigned int length,
+								bool partial)
+{
+	if (length == 0)
+		{
+			if (stack == NULL)
+				return true;
+			else if (partial)
+				return true;
+			else
+				return false;
+		}
+
+	if (stack == NULL)
+		return false;
+
+	if (stack->type == expectation[0])
+		return stateStackMatch (stack->next, expectation + 1, length - 1, partial);
+	else
+		return false;
+}
+
+static bool scalarNeq (yaml_token_t *token, unsigned int len, const char* val)
+{
+	if ((token->data.scalar.length == len)
+		&& (strncmp (val, (char *)token->data.scalar.value, len) == 0))
+		return true;
+	else
+		return false;
+}
+
+static void	ansiblePlaybookPlayStateMachine (struct ansiblePlaybookState *state,
+											 yaml_token_t *token)
+{
+	yaml_token_type_t play_expect[] = {
+		YAML_BLOCK_MAPPING_START_TOKEN,
+		YAML_BLOCK_SEQUENCE_START_TOKEN,
+	};
+
+	switch (token->type)
+		{
+		case YAML_KEY_TOKEN:
+			if (stateStackMatch (state->type_stack,
+								 play_expect, ARRAY_SIZE (play_expect),
+								 false))
+				state->play_detection_state = DSTAT_PLAY_NAME_KEY;
+			else
+				state->play_detection_state = DSTAT_PLAY_NAME_INITIAL;
+			break;
+		case YAML_SCALAR_TOKEN:
+			if ((state->play_detection_state == DSTAT_PLAY_NAME_KEY)
+				&& scalarNeq (token, 4, "name"))
+				state->play_detection_state = DSTAT_PLAY_NAME_KEY_SCALAR;
+			else if (state->play_detection_state == DSTAT_PLAY_NAME_VALUE)
+				{
+					tagEntryInfo tag;
+					initTagEntry (&tag, (char *)token->data.scalar.value,
+								  AnsiblePlaybookKinds + K_PLAY);
+					attachYamlPosition (&tag, token, false);
+
+					Assert (state->type_stack->associatedCorkIndex == CORK_NIL);
+					state->type_stack->associatedCorkIndex = makeTagEntry (&tag);
+					state->play_detection_state = DSTAT_PLAY_NAME_INITIAL;
+				}
+			else
+				state->play_detection_state = DSTAT_PLAY_NAME_INITIAL;
+			break;
+		case YAML_VALUE_TOKEN:
+			if (state->play_detection_state == DSTAT_PLAY_NAME_KEY_SCALAR)
+				state->play_detection_state = DSTAT_PLAY_NAME_VALUE;
+			else
+				state->play_detection_state = DSTAT_PLAY_NAME_INITIAL;
+			break;
+		default:
+			state->play_detection_state = DSTAT_PLAY_NAME_INITIAL;
+			break;
+		}
+}
+
+static void ansiblePlaybook (yaml_token_t *token, void *data)
+{
+	struct ansiblePlaybookState *state = data;
+
+	if (token->type == YAML_BLOCK_SEQUENCE_START_TOKEN
+		|| token->type == YAML_BLOCK_MAPPING_START_TOKEN)
+		pushBlockType (state, token->type);
+
+	ansiblePlaybookPlayStateMachine (state, token);
+
+	if (token->type == YAML_BLOCK_END_TOKEN)
+		popBlockType (state, token);
+	else if (token->type == YAML_STREAM_END_TOKEN)
+		popAllBlockType (data, token);
+}
+
+static void *ansiblePlaybookPrepare(void)
+{
+	static struct ansiblePlaybookState state = {
+		.type_stack = NULL,
+		.play_detection_state = DSTAT_PLAY_NAME_INITIAL,
+	};
```

Comment:
Unless I'm mistaken, this won't reset the state when parsing second and next files, won't that be a problem?